fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/RankedSetIndexHelper.java

```diff
@@ -233,38 +244,41 @@ private static Number extractRank(int groupPrefixSize, @Nullable Tuple maybeRank
                                                           @Nonnull Tuple valueKey,
                                                           @Nonnull Tuple scoreKey,
                                                           boolean remove) {
-        final RankedSet rankedSet = new InstrumentedRankedSet(state, rankSubspace, hashFunction, nlevels);
+        final RankedSet rankedSet = new InstrumentedRankedSet(state, rankSubspace, config);
         final byte[] score = scoreKey.pack();
         CompletableFuture<Void> result = init(state, rankedSet).thenCompose(v -> {
             if (remove) {
-                // If no one else has this score, remove from ranked set.
-                return state.transaction.getRange(state.indexSubspace.range(valueKey)).iterator().onHasNext().thenCompose(hasNext -> {
-                    if (hasNext) {
-                        return AsyncUtil.DONE;
-                    } else {
-                        return rankedSet.remove(state.transaction, score).thenApply(exists -> {
-                            // It is okay if the score isn't in the ranked set yet if the index is
-                            // write only because this means that the score just hasn't yet
-                            // been added by some record. We still want the conflict ranges, though.
-                            if (!exists && !state.store.isIndexWriteOnly(state.index)) {
-                                throw new RecordCoreException("Score was not present in ranked set.");
-                            }
-                            return null;
-                        });
-                    }
-                });
+                if (config.isCountDuplicates()) {
+                    // Decrement count and possibly remove.
+                    return removeFromRankedSet(state, rankedSet, score);
+                } else {
+                    // If no one else has this score, remove from ranked set.
+                    return state.transaction.getRange(state.indexSubspace.range(valueKey)).iterator().onHasNext()
+                            .thenCompose(hasNext -> hasNext ? AsyncUtil.DONE : removeFromRankedSet(state, rankedSet, score));
+                }
             } else {
                 return rankedSet.add(state.transaction, score).thenApply(added -> null);
             }
         });
         return state.store.instrument(Events.RANKED_SET_UPDATE, result);
     }
 
+    private static CompletableFuture<Void> removeFromRankedSet(@Nonnull IndexMaintainerState state, @Nonnull RankedSet rankedSet, @Nonnull byte[] score) {
+        return rankedSet.remove(state.transaction, score).thenApply(exists -> {
+            // It is okay if the score isn't in the ranked set yet if the index is
+            // write only because this means that the score just hasn't yet
+            // been added by some record. We still want the conflict ranges, though.
```

Comment:
I, um, am not sure this comment is true, or at least doesn't seem right given one of the comments I made in the `remove` function for `RankedSet`. On the other hand, we do test write only indexes as part of index builds, so I'm not entirely sure. (Understanding that this is a pre-existing comment that I probably wrote.)

Reply:
I somehow missed the part of the code in `RankedSet` that checks to make sure that the score isn't zero before decrementing. I am now more confident in this comment in at least the case of not duplicate tracking (so that's good).

I'm trying to think, however, if there are pathologies in the count duplicates case. I think the answer is "yes" because adding to a ranked set index with duplicate counting isn't idempotent. In particular:

1. Imagine having two records with a rank index on field `foo`, both of which have the value `foo = bar`.
1. An index build indexes the first record. Now the count corresponding to `bar` in the `foo` rank index is 1.
1. Someone removes the second record. Now the count corresponding to `bar` in the `foo` rank index is 0 (i.e., it is removed from the ranked set).
1. The index build completes, never having indexed the second record.

So now `bar` has an incorrect rank. This is somewhat of a worst case in that `bar` is missing entirely from the ranked set, but I think one can also be in a position where the rank is wrong, but not removed entirely.

I think this can be fixed by checking if the record is in the built range and only updating the ranked set if it is. It also seems to me like this would be safe for the regular ranked set, though perhaps would induce more conflicts.

Reply:
That would be accomplished by having `isIdempotent` check the option, correct? There is no need for a new instance of such built range checking code.

Reply:
Yeah, I believe so--that no new mechanism is necessary because what I sketched out is already done by `isIdempotent`.